In [1]:
# execute to import notebook styling for tables and width etc.
from IPython.core.display import HTML
import urllib.request
response = urllib.request.urlopen('https://raw.githubusercontent.com/DataScienceUWL/DS775v2/master/ds755.css')
HTML(response.read().decode("utf-8"));

<font size=18>Lesson 10 - Self-Assessment Solutions</font>

# <font color = "blue"> Self-Assessment: Load and Display - Solution</font>

In [76]:
import pandas as pd
import numpy as np

ted = pd.read_csv('./data/ted-talks/ted_main.csv')
ted.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  published_date  \
0      Ken Robinson: Do schools kill creativity?            1      1151367060   
1           Al Gore: Averting the climate crisis            1      1151367060   
2                  David Pogue: Simplicity sells            1      1151367060   
3             Majora Carter: Greening the ghetto            1      1151367060   
4  Hans Rosling: The best stats you've ever seen            1      1151440680   

                                             ratings  \
0  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869

# <font color = "blue"> Self-Assessment: Pandas - Solution </font>

In [2]:
ted.shape

(2550, 17)

There are 2550 TED talks in this data frame.

# <font color = "blue"> Self-Assessment: Prerequisites - Solution </font> 

In [40]:
#Calculate the number of views for the 10th percentile 
m = ted['views'].quantile(0.10)

#Only consider talks of at least 5 minutes
q_talks = ted[(ted['duration'] >= 300)]

#Only consider talks in the top 90%
q_talks = q_talks[q_talks['views'] >= m]

#Inspect the number of talks that made the cut
q_talks.shape

(2151, 17)

# <font color = "blue"> Self-Assessment: Compute a Metric, Sort and Print - Solution </font>  

In [41]:
#create the metric of the comments to views ratio
q_talks['comments_per_1000views']=1000*q_talks['comments']/q_talks['views']

#Sort talks in descending order of the ratio of views to comments
q_talks = q_talks.sort_values('comments_per_1000views', ascending=False)

#Print the top 10 talks
q_talks[['description', 'main_speaker', 'comments_per_1000views']].head(10)


description  \
803   David Bismark demos a new system for voting th...   
96    Richard Dawkins urges all atheists to openly s...   
694   Filmmaker Sharmeen Obaid-Chinoy takes on a ter...   
954   Janet Echelman found her true voice as an arti...   
840   Lesley Hazleton sat down one day to read the K...   
1787  Our consciousness is a fundamental aspect of o...   
661   Vaccine-autism claims, "Frankenfood" bans, the...   
443   At TED2009, Al Gore presents updated slides fr...   
732   In "THE 99," Naif Al-Mutawa's new generation o...   
1148  What does environmental devastation actually l...   

               main_speaker  comments_per_1000views  
803           David Bismark                1.534355  
96          Richard Dawkins                1.463841  
694   Sharmeen Obaid-Chinoy                1.420683  
954          Janet Echelman                1.359572  
840         Lesley Hazleton                1.285149  
1787         David Chalmers                1.235918  
661         Michael Specter                1.235704  
443                 Al Gore                1.091008  
732          Naif Al-Mutawa                0.981252  
1148             Garth Lenz                0.977574

# <font color = "blue"> Self-Assessment: Dealing with Dates - Solution </font>

In [42]:
#Convert release_date into pandas datetime format
ted['published_date'] = pd.to_datetime(ted['published_date'],
                                       errors='coerce', unit='s', origin='unix')

#Extract year from the datetime
ted['published_year'] = ted['published_date'].apply(lambda x: str(x).split('-')[0] if x != np.nan else np.nan)

#Helper function to convert NaT to 0 and all other years to integers.
def convert_int(x):
    try:
        return int(x)
    except:
        return 0

#Apply convert_int to the year feature
ted['published_year'] = ted['published_year'].apply(convert_int)
    
ted.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  \
0      Ken Robinson: Do schools kill creativity?            1   
1           Al Gore: Averting the climate crisis            1   
2                  David Pogue: Simplicity sells            1   
3             Majora Carter: Greening the ghetto            1   
4  Hans Rosling: The best stats you've ever seen            1   

       published_date                                            ratings  \
0 2006-06-27 00:11:00  [{'id': 7, 'name': 'Funny', 'count': 19645}, {...   
1 2006-06-27 00:11:00  [{'id': 7, 'name': 'Funny', 'count': 544}, {'i...   
2 2006-06-27 00:11:00  [{'id': 7, 'name': 'Funny', 'count': 964}, {'i...   
3 2006-06-27 00:11:00  [{'id': 3, 'name': 'Courageous', 'count': 760}...   
4 2006-06-27 20:38:00  [{'id': 9, 'name': 'Ingenious', 'count': 3202}...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  published_year  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110            2006  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520            2006  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292            2006  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550            2006  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869            2006

# <font color = "blue"> Self-Assessment: Stringified Dictionaries - Solution </font>

In [43]:
#Import the literal_eval function from ast
from ast import literal_eval

#Convert all NaN into stringified empty lists
ted['ratings'] = ted['ratings'].fillna('[]')

#Apply literal_eval to convert stringified empty lists to the list object
ted['ratings'] = ted['ratings'].apply(literal_eval)

#Convert list of dictionaries to a list of strings
ted['ratings'] = ted['ratings'].apply(lambda x: [i['name'].lower() for i in x] if isinstance(x, list) else [])

#See how 'ratings' has changed?
ted.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
1       265  With the same humor and humanity he exuded in ...       977   
2       124  New York Times columnist David Pogue takes aim...      1286   
3       200  In an emotionally charged talk, MacArthur-winn...      1116   
4       593  You've never seen data presented like this. Wi...      1190   

     event   film_date  languages   main_speaker  \
0  TED2006  1140825600         60   Ken Robinson   
1  TED2006  1140825600         43        Al Gore   
2  TED2006  1140739200         26    David Pogue   
3  TED2006  1140912000         35  Majora Carter   
4  TED2006  1140566400         48   Hans Rosling   

                                            name  num_speaker  \
0      Ken Robinson: Do schools kill creativity?            1   
1           Al Gore: Averting the climate crisis            1   
2                  David Pogue: Simplicity sells            1   
3             Majora Carter: Greening the ghetto            1   
4  Hans Rosling: The best stats you've ever seen            1   

       published_date                                            ratings  \
0 2006-06-27 00:11:00  [funny, beautiful, ingenious, courageous, long...   
1 2006-06-27 00:11:00  [funny, courageous, confusing, beautiful, unco...   
2 2006-06-27 00:11:00  [funny, courageous, ingenious, beautiful, unco...   
3 2006-06-27 00:11:00  [courageous, beautiful, confusing, funny, inge...   
4 2006-06-27 20:38:00  [ingenious, funny, beautiful, courageous, long...   

                                       related_talks  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...   
1  [{'id': 243, 'hero': 'https://pe.tedcdn.com/im...   
2  [{'id': 1725, 'hero': 'https://pe.tedcdn.com/i...   
3  [{'id': 1041, 'hero': 'https://pe.tedcdn.com/i...   
4  [{'id': 2056, 'hero': 'https://pe.tedcdn.com/i...   

                     speaker_occupation  \
0                       Author/educator   
1                      Climate advocate   
2                  Technology columnist   
3    Activist for environmental justice   
4  Global health expert; data visionary   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
1  ['alternative energy', 'cars', 'climate change...   
2  ['computers', 'entertainment', 'interface desi...   
3  ['MacArthur grant', 'activism', 'business', 'c...   
4  ['Africa', 'Asia', 'Google', 'demo', 'economic...   

                             title  \
0      Do schools kill creativity?   
1      Averting the climate crisis   
2                 Simplicity sells   
3              Greening the ghetto   
4  The best stats you've ever seen   

                                                 url     views  published_year  
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110            2006  
1  https://www.ted.com/talks/al_gore_on_averting_...   3200520            2006  
2  https://www.ted.com/talks/david_pogue_says_sim...   1636292            2006  
3  https://www.ted.com/talks/majora_carter_s_tale...   1697550            2006  
4  https://www.ted.com/talks/hans_rosling_shows_t...  12005869            2006

In [44]:
#Create a new feature by exploding word ratings
s = ted.apply(lambda x: pd.Series(x['ratings']),axis=1).stack().reset_index(level=1, drop=True)

#Name the new feature as 'word_ratings'
s.name = 'word_ratings'

#Create a new dataframe gen_ted which by dropping the old 'ratings' feature and adding the new 'word_ratings'.
gen_ted = ted.drop('ratings', axis=1).join(s)

#scorll to the right to see the resulting 'word_ratings' variable
gen_ted.head()

comments                                        description  duration  \
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   
0      4553  Sir Ken Robinson makes an entertaining and pro...      1164   

     event   film_date  languages  main_speaker  \
0  TED2006  1140825600         60  Ken Robinson   
0  TED2006  1140825600         60  Ken Robinson   
0  TED2006  1140825600         60  Ken Robinson   
0  TED2006  1140825600         60  Ken Robinson   
0  TED2006  1140825600         60  Ken Robinson   

                                        name  num_speaker      published_date  \
0  Ken Robinson: Do schools kill creativity?            1 2006-06-27 00:11:00   
0  Ken Robinson: Do schools kill creativity?            1 2006-06-27 00:11:00   
0  Ken Robinson: Do schools kill creativity?            1 2006-06-27 00:11:00   
0  Ken Robinson: Do schools kill creativity?            1 2006-06-27 00:11:00   
0  Ken Robinson: Do schools kill creativity?            1 2006-06-27 00:11:00   

                                       related_talks speaker_occupation  \
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...    Author/educator   
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...    Author/educator   
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...    Author/educator   
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...    Author/educator   
0  [{'id': 865, 'hero': 'https://pe.tedcdn.com/im...    Author/educator   

                                                tags  \
0  ['children', 'creativity', 'culture', 'dance',...   
0  ['children', 'creativity', 'culture', 'dance',...   
0  ['children', 'creativity', 'culture', 'dance',...   
0  ['children', 'creativity', 'culture', 'dance',...   
0  ['children', 'creativity', 'culture', 'dance',...   

                         title  \
0  Do schools kill creativity?   
0  Do schools kill creativity?   
0  Do schools kill creativity?   
0  Do schools kill creativity?   
0  Do schools kill creativity?   

                                                 url     views  \
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110   
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110   
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110   
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110   
0  https://www.ted.com/talks/ken_robinson_says_sc...  47227110   

   published_year word_ratings  
0            2006        funny  
0            2006    beautiful  
0            2006    ingenious  
0            2006   courageous  
0            2006   longwinded

# <font color = "blue"> Self-Assessment: Create the Knowledge-Based Recommender - Solution </font>

In [58]:
def build_chart(gen_ted, percentile=0.1):
    #Show user the list of word ratings to choose from
    print(gen_ted['word_ratings'].unique())
    
    #Ask for preferred word rating
    print("Select a descriptive word from the list above for the 'word rating'")
    rating = input()
    
    #Ask for lower limit of film year
    print("Input earliest year published (2006 to 2017)")
    low_year = int(input())
    
    #Ask for upper limit of film year
    print("Input latest year published(2006 to 2017)")
    high_year = int(input())
    
    
    #Define a new talks variable to store the preferred talks. 
    #Copy the contents of gen_ted to talks
    talks = gen_ted.copy()
    
    #Filter based on the condition
    talks = talks[(talks['word_ratings'] == rating) & 
                    (talks['published_year'] >= low_year) & 
                    (talks['published_year'] <= high_year)]
    
    #Calculate the number of views for the 10th percentile 
    m = ted['views'].quantile(percentile)

    #Only consider movies that have higher than m votes. Save this in a new dataframe q_movies
    q_talks = talks.copy().loc[talks['views'] >= m]
    
    #create the metric of the comments to views ratio
    q_talks['comments_per_1000views']=1000*q_talks['comments']/q_talks['views']

    #Sort talks in descending order of the ratio of views to comments
    q_talks = q_talks.sort_values('comments_per_1000views', ascending=False)
    
    return q_talks

In [80]:
#Generate the chart for top talks for these user preferences and display top 5.
gen_ted_final = build_chart(gen_ted).head(5)

gen_ted_final[['main_speaker','name','published_year','comments_per_1000views']]

['funny' 'beautiful' 'ingenious' 'courageous' 'longwinded' 'confusing'
 'informative' 'fascinating' 'unconvincing' 'persuasive' 'jaw-dropping'
 'ok' 'obnoxious' 'inspiring']
Select a descriptive word from the list above for the 'word rating'
obnoxious
Input earliest year published (2006 to 2017)
2009
Input latest year published(2006 to 2017)
2014


main_speaker  \
803           David Bismark   
694   Sharmeen Obaid-Chinoy   
954          Janet Echelman   
840         Lesley Hazleton   
1787         David Chalmers   

                                                   name  published_year  \
803               David Bismark: E-voting without fraud            2010   
694   Sharmeen Obaid-Chinoy: Inside a school for sui...            2010   
954        Janet Echelman: Taking imagination seriously            2011   
840               Lesley Hazleton: On reading the Koran            2011   
1787  David Chalmers: How do you explain consciousness?            2014   

      comments_per_1000views  
803                 1.534355  
694                 1.420683  
954                 1.359572  
840                 1.285149  
1787                1.235918

# <font color = "blue"> Self-Assessment: TF-IDF Vectors - Solution</font>

In [62]:
#Import TfIdfVectorizer from the scikit-learn library
from sklearn.feature_extraction.text import TfidfVectorizer

#Define a TF-IDF Vectorizer Object. Remove all english stopwords
tfidf = TfidfVectorizer(stop_words='english')

#Replace NaN with an empty string
ted['description'] = ted['description'].fillna('')

#Construct the required TF-IDF matrix by applying the fit_transform method on the overview feature
tfidf_matrix = tfidf.fit_transform(ted['description'])

#Output the shape of tfidf_matrix
tfidf_matrix.shape

(2550, 15162)

# <font color = "blue"> Self-Assessment: Create the Content-Based Recommender Based on Cosine Similarity - Solution </font>

In [74]:
# Import linear_kernel to compute the dot product
from sklearn.metrics.pairwise import linear_kernel

# Compute the cosine similarity matrix
cosine_sim = linear_kernel(tfidf_matrix, tfidf_matrix)

#Construct a reverse mapping of indices and talk names, and drop duplicate names, if any
indices = pd.Series(ted.index, index=ted['name']).drop_duplicates()

In [85]:
# Function that takes in talk title as input and gives recommendations 
def content_recommender(name, cosine_sim=cosine_sim, df=ted, indices=indices):
    # Obtain the index of the talks that matches the title
    idx = indices[name]

    # Get the pairwsie similarity scores of all talks with that name
    # And convert it into a list of tuples as described above
    sim_scores = list(enumerate(cosine_sim[idx]))

    # Sort the talks based on the cosine similarity scores
    sim_scores = sorted(sim_scores, key=lambda x: x[1], reverse=True)

    # Get the scores of the 10 most similar talks. Ignore the first talk.
    sim_scores = sim_scores[1:11]

    # Get the movie indices
    movie_indices = [i[0] for i in sim_scores]

    # Return the top 10 most similar movies
    return df['name'].iloc[movie_indices]

In [86]:
#Get recommendations for Tyler Cowen: Be suspicious of simple stories
content_recommender('Tyler Cowen: Be suspicious of simple stories')

1910    Tom Wujec: Got a wicked problem? First, tell m...
451     Dan Ariely: Are we in control of our own decis...
1115         Mikko Hypponen: Three types of online attack
1405           Ronny Edry: Israel and Iran: A love story?
2361     Sisonke Msimang: If a story moves you, act on it
406                      Dan Ariely: Our buggy moral code
775         Julian Treasure: Shh! Sound health in 8 steps
335     Samantha Power: A complicated hero in the war ...
145                  John Maeda: Designing for simplicity
757     His Holiness the Karmapa: The technology of th...
Name: name, dtype: object